pkg/licensing/normalize.go

```diff
@@ -175,16 +587,84 @@ var pythonLicenseExceptions = map[string]string{
 // 'GPL-1+ or Artistic, and BSD-4-clause-POWERDOG' => {"GPL-1+", "Artistic", "BSD-4-clause-POWERDOG"}
 // 'BSD 3-Clause License or Apache License, Version 2.0' => {"BSD 3-Clause License", "Apache License, Version 2.0"}
-// var LicenseSplitRegexp = regexp.MustCompile("(,?[_ ]+or[_ ]+)|(,?[_ ]+and[_ ])|(,[ ]*)")
-
 var licenseSplitRegexp = regexp.MustCompile("(,?[_ ]+(?:or|and)[_ ]+)|(,[ ]*)")
 
-func Normalize(name string) string {
+// version number match
+var versionRegexpString = "([A-UW-Z)]{2,})( LICENSE)?\\s*[,(-]?\\s*(V|V\\.|VERSION|VERSION-|-)?\\s*([1-9](\\.\\d)*)[)]?"
+
+// case insensitive version match anywhere in string
+var versionRegexp = regexp.MustCompile("(?i)" + versionRegexpString)
+
+// version suffix match
+var versionSuffixRegexp = regexp.MustCompile(versionRegexpString + "$")
+
+// suffixes from https://spdx.dev/learn/handling-license-info/
+var onlySuffixes = [2]string{"-ONLY", " ONLY"}
+var plusSuffixes = [3]string{"+", "-OR-LATER", " OR LATER"}
+
+func standardizeKeyAndSuffix(name string) expression.SimpleExpr {
+	// Standardize space, including newline
+	name = strings.Join(strings.Fields(name), " ")
 	name = strings.TrimSpace(name)
-	if l, ok := mapping[strings.ToUpper(name)]; ok {
-		return l
+	name = strings.ToUpper(name)
+	// Do not perform any further normalization for URLs
+	if strings.HasPrefix(name, "HTTP") {
+		return expression.SimpleExpr{License: name, HasPlus: false}
+	}
+	name = strings.ReplaceAll(name, "LICENCE", "LICENSE")
+	name = strings.TrimPrefix(name, "THE ")
+	name = strings.TrimSuffix(name, " LICENSE")
+	name = strings.TrimSuffix(name, " LICENSED")
+	name = strings.TrimSuffix(name, "-LICENSE")
+	name = strings.TrimSuffix(name, "-LICENSED")
+	// Remove License and Licensed suffixes except for licenses already containing those suffixes such as Unlicense
+	if name != "UNLICENSE" {
+		name = strings.TrimSuffix(name, "LICENSE")
+	}
+	if name != "UNLICENSED" {
+		name = strings.TrimSuffix(name, "LICENSED")
+	}
+	hasPlus := false
+	for _, s := range plusSuffixes {
+		if strings.HasSuffix(name, s) {
+			name = strings.TrimSuffix(name, s)
+			hasPlus = true
+		}
+	}
+	for _, s := range onlySuffixes {
+		name = strings.TrimSuffix(name, s)
 	}
-	return name
+	name = versionSuffixRegexp.ReplaceAllString(name, "$1-$4")
+	return expression.SimpleExpr{License: name, HasPlus: hasPlus}
+}
+
+// Returns invalid keys in a map between license strings and normalized licenses.
+// The map keys must be standardized (uppercase, no common suffixes, etc.).
+// A nil argument will return invalid keys in the default mapping.
+func InvalidMappingKeys(licenseToNormalized map[string]expression.SimpleExpr) []string {
+	if licenseToNormalized == nil {
+		licenseToNormalized = mapping
+	}
+	var invalid []string
+	for key := range licenseToNormalized {
+		standardized := standardizeKeyAndSuffix(key)
+		if standardized.License != key {
+			invalid = append(invalid, key)
+		}
+	}
+	return invalid
+}
```

Comment:
What if we move this function to the test.
To get `mapping`, we can create a function like [GetBuiltinRules()](https://github.com/aquasecurity/trivy/blob/70245721372720027b7089bd61c693df48add865/pkg/fanal/secret/builtin-rules.go#L86-L88) for secrets.

Reply:
normalize_test is now in package licensing to use internal mapping and functions.